File: 11-20/17. Letter Combinations of a Phone Number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
    if (digits.length == 0) return [];
    let map = [
        [],
        [],
        ["a", "b", "c"],
        ["d", "e", "f"],
        ["g", "h", "i"],
        ["j", "k", "l"],
        ["m", "n", "o"],
        ["p", "q", "r", "s"],
        ["t", "u", "v"],
        ["w", "x", "y", "z"],
    ];

    let ans = [];
    let len = digits.length;
    //DFS
    const iter = (str, index) => {
        if (index == len) {
            ans.push(str);
            return;
        }

        let current = digits[index] - "0";
        let chars = map[current];
        for (let i = 0; i < chars.length; i++) {
            iter(str + chars[i], index + 1);
        }
    };

    iter("", 0);
    return ans;
};

letterCombinations("23");

File: 11-20/17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not len(digits):
            return []
        pMap = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"],
        }

        res = []
        string = ""

        def iter(i):
            nonlocal string
            if i == len(digits):
                res.append(string)
                return
            chars = pMap[digits[i]]
            for c in chars:
                string += c
                iter(i + 1)
                string = string[:-1]

        iter(0)
        return res
